src/main/java/com/cburch/logisim/std/ttl/Ttl74381.java

```diff
@@ -0,0 +1,265 @@
+
+  public static final byte S0 = 5;
+  public static final byte S1 = 6;
+  public static final byte S2 = 7;
+
+  public static final byte Ci = 15;
+
+  // Outputs
+  public static final byte P = 14;
+  public static final byte G = 13;
+
+  // Power supply
+  public static final byte GND = 10;
+  public static final byte VCC = 20;
+
+  private InstanceState _state;
+
+  public Ttl74381() {
+    super(
+            _ID,
+            (byte) 20,
+            new byte[] {
+              F0, F1, F2, F3, P, G
+            },
+            new String[] {
+              "A1", "B1", "A0", "B0", "S0", "S1", "S2", "F0", "F1",
+              "F2", "F3", "Gn", "Pn", "Ci", "B3", "A3", "B2", "A2"
+            },
+            null);
+  }
+
+  @Override
+  public void paintInternal(InstancePainter painter, int x, int y, int height, boolean up) {
+    // As tooltips can be longer than what can fit as pin name while painting IC internals,
+    // we need to shorten it first to up to 4 characters to keep the diagram readable.
+    final var label_len_max = 4;
+    final var names = new ArrayList<String>();
+    for (final var name : portNames) {
+      final var tmp = name.split("\\s+");
+      names.add((tmp[0].length() <= label_len_max) ? tmp[0] : tmp[0].substring(0, label_len_max));
+    }
+    super.paintBase(painter, true, false);
+    Drawgates.paintPortNames(painter, x, y, height, names.toArray(new String[0]));
+  }
+
+  /** IC pin indices are datasheet based (1-indexed), but ports are 0-indexed
+   *
+   * @param dsPinNr datasheet pin number
+   * @return port number
+   */
+  protected byte pinNrToPortNr(byte dsPinNr) {
+    return (byte) ((dsPinNr <= GND) ? dsPinNr - 1 : dsPinNr - 2);
+  }
+
+  /** Gets the current state of the specified pin
+   *
+   * @param dsPinNr datasheet pin number
+   * @return true if the specified pin has a logic high level
+   */
+  private boolean getPort(byte dsPinNr) {
+    return _state.getPortValue(pinNrToPortNr(dsPinNr)) == Value.TRUE;
+  }
+
+  /** Sets the specified pin to the specified level
+   *
+   * @param dsPinNr datasheet pin number
+   * @param b the logic level for the pin
+   */
+  private void setPort(byte dsPinNr, boolean b) {
+    _state.setPort(pinNrToPortNr(dsPinNr), b ? Value.TRUE : Value.FALSE, DELAY);
+  }
+
+  /**
+   * Implements a "PAL32L1" device i.e. a programmable array logic device with a virtually
+   * unlimited sum of products with up to 32 inputs for each product.
+   *
+   * @param in the inputs
+   * @param products list of products, if bit n of a product is a one, then input n is part of that product
+   * @return the active-low sum of products
+   */
+  private boolean pal32L1(ArrayList<Boolean> in, int[] products) {
+    var or = false;
+
+    for (var i = 0; i < products.length && !or; i++) {
+      var product = products[i];
+      var and = true;
+
+      for (var j = 0; j < 32 && and; j++) {
+        if ((product & (1 << j)) != 0) {
+          and = in.get(j);

```

Comment:
shouldn't you `break` here too?

Reply:
The check on "and" in the for statement takes care of early exit.

Reply:
Oh, indeed. But I'd insist on using `break` to jump out, for better code readability. Thoughts?

Reply:
For me, it is clear enough, but maybe adding a small comment would help.

Reply:
Added comments, both at the product term and the sum term assignments

Reply:
@MarcinOrlowski: Could you please mark it as resolved?